### PR TITLE
ORDER BY ColumnNumber with Collations

### PR DESCRIPTION
**Pull request by tiagokepe**
Hello all,

This PR enables ORDER BY ColumnNumber with collations like `SELECT 'a' ORDER BY 1 COLLATE NOCASE`.

### CODE REVIEW

File: src/planner/expression_binder/order_binder.cpp [hunk 1]
```diff
@@ -100,6 +101,24 @@ unique_ptr<Expression> OrderBinder::Bind(unique_ptr<ParsedExpression> expr) {
 	case ExpressionClass::PARAMETER: {
 		throw ParameterNotAllowedException("Parameter not supported in ORDER BY clause");
 	}
+	case ExpressionClass::COLLATE: {
+		auto &collation = expr->Cast<CollateExpression>();
+		if (collation.child->expression_class == ExpressionClass::CONSTANT) {
+			auto &constant = collation.child->Cast<ConstantExpression>();
+			auto index = (idx_t)constant.value.GetValue<int64_t>() - 1;
```
Comment (Tishj) [inline src/planner/expression_binder/order_binder.cpp:108]: I believe this needs a `NumericCast`, and we should add a test with a negative value

File: test/sql/collate/test_collate_orderby_column_number.test [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+a
+
+statement ok
+CREATE TABLE collate_test(s VARCHAR)
+
+statement ok
+INSERT INTO collate_test VALUES ('ã'),('B'),('a'),('A');
+
+query I
+SELECT s FROM collate_test ORDER BY 1 COLLATE NOCASE;
```
Comment (Mytherin) [inline test/sql/collate/test_collate_orderby_column_number.test:25]: In addition to the tests @Tishj requested could we add some tests with more complex expressions as well? i.e.

```sql
SELECT CONCAT(s, s) FROM collate_test ORDER BY 1 COLLATE NOCASE;
SELECT (SELECT s) FROM collate_test ORDER BY 1 COLLATE NOCASE;
SELECT CONCAT(s, s) AS concat, concat FROM collate_test ORDER BY 2 COLLATE NOCASE;
SELECT CASE WHEN s[1]='a' THEN s ELSE NULL END FROM collate_test ORDER BY 1 COLLATE NOCASE;
SELECT collate_test.s FROM collate_test ORDER BY 1 COLLATE NOCASE;
```

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good - one comment:

File: src/planner/expression_binder/order_binder.cpp [hunk 3]
```diff
@@ -100,6 +101,31 @@ unique_ptr<Expression> OrderBinder::Bind(unique_ptr<ParsedExpression> expr) {
...
+		if (collation.child->expression_class == ExpressionClass::CONSTANT) {
+			auto &constant = collation.child->Cast<ConstantExpression>();
+			auto index = NumericCast<idx_t>(constant.value.GetValue<idx_t>()) - 1;
+			if (index >= extra_list->size()) {
+				throw BinderException("ORDER term out of range - should be between 1 and %lld", (idx_t)max_count);
+			}
+			auto &sel_entry = extra_list->at(index);
+			if (!sel_entry->alias.empty()) {
+				vector<string> column_names = {sel_entry->alias};
+				auto colref = make_uniq<ColumnRefExpression>(std::move(column_names));
```
Comment (Mytherin) [inline src/planner/expression_binder/order_binder.cpp:115]: I don't quite follow what the goal of this code path is, e.g. it seems like this would not work:

```sql
SELECT 'a' AS my_alias ORDER BY 1 COLLATE NOCASE;
```

Perhaps we can just remove this code path and always use copy?

File: src/planner/expression_binder/order_binder.cpp [hunk 4]
```diff
@@ -100,6 +101,31 @@ unique_ptr<Expression> OrderBinder::Bind(unique_ptr<ParsedExpression> expr) {
...
+				colref->query_location = sel_entry->query_location;
+				collation.child = std::move(colref);
+			} else { // no alias, e.g.: SELECT 'a' ORDER BY 1 COLLATE NOCASE
+				if (sel_entry->expression_class == ExpressionClass::SUBQUERY) {
```
Comment (Mytherin) [inline src/planner/expression_binder/order_binder.cpp:119]: This should use `HasSubquery()`, otherwise e.g. this will result in an error:

```sql
SELECT concat('a', (SELECT s)) FROM collate_test ORDER BY 1 COLLATE NOCASE;
```

**Review by Mytherin — COMMENTED** [review]
Thanks for the updates - some more comments:

File: src/planner/expression_binder/order_binder.cpp (hunk 4 above)
Comment (tiagokepe) [reply src/planner/expression_binder/order_binder.cpp]: Thanks @Mytherin, I've applied all suggestions and added extra test cases.

**Comment by Mytherin** [comment]
Thanks! LGTM
